### PR TITLE
rt: simplify coop implementation

### PR DESCRIPTION
**Pull request by carllerche**
Simplifies coop implementation. Prunes unused code, create a `Budget`
type to track the current budget.

### CODE REVIEW

File: tokio/src/sync/batch_semaphore.rs [hunk 1]
```diff
@@ -386,7 +386,11 @@ impl Future for Acquire<'_> {
     type Output = Result<(), AcquireError>;
 
     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // First, ensure the current task has enough budget to proceed.
+        ready!(crate::coop::poll_proceed(cx));
```
Comment (hawkw) [inline tokio/src/sync/batch_semaphore.rs:390]: :+1: makes sense to just put this here.

**Comment by carllerche** [comment]
This is in preparation for a bug fix. The following test fails:

```rust
#[test]
#[ignore]
fn coop_and_block_in_place() {
    use tokio::sync::mpsc;

    let mut rt = rt();

    rt.block_on(async move {
        let (mut tx, mut rx) = mpsc::channel(1024);

        // Fill the channel
        for _ in 0..1024 {
            tx.send(()).await.unwrap();
        }

        drop(tx);

        tokio::spawn(async move {
            // Block in place without doing anything
            tokio::task::block_in_place(|| {});

            // Receive all the values, this should trigger a `Pending` as the
            // coop limit will be reached.
            poll_fn(|cx| {
                while let Poll::Ready(v) = {
                    tokio::pin! {
                        let fut = rx.recv();
                    }

                    Pin::new(&mut fut).poll(cx)
                } {
                    if v.is_none() {
                        panic!("did not yield");
                    }
                }

                Poll::Ready(())
            })
            .await
        })
        .await
        .unwrap();
    });
}
```

File: tokio/src/coop.rs [hunk 2]
```diff
@@ -29,384 +31,188 @@
...
+            }
+        } else {
+            true
+        }
+    }
 }
 
-/// Run the given closure with a cooperative task budget.
-///
-/// Enabling budgeting when it is already enabled is a no-op.
```
Comment (hawkw) [inline tokio/src/coop.rs:73]: Was this behavior being used somewhere previously?

Comment (carllerche) [reply tokio/src/coop.rs:73]: no tests as far as I can see. Based on LocalSet, it seems like it is the desired behavior.

File: tokio/src/macros/cfg.rs [hunk 3]
```diff
@@ -363,3 +363,23 @@ macro_rules! cfg_unstable {
...
+                    feature = "fs",
+                    feature = "io-driver",
+                    feature = "io-std",
+                    feature = "process",
+                    feature = "rt-core",
+                    feature = "sync",
+                    feature = "stream",
+                    feature = "time"
+                    ))]
+            $item
```
Comment (carllerche) [inline tokio/src/macros/cfg.rs:382]: This is a bit unfortunate. Odds are we probably should just silence warnings for coop, but doing that makes it easy to miss dead code during refactors...

File: tokio/src/coop.rs [hunk 4]
```diff
@@ -29,384 +31,197 @@
...
-/// Now, even if `big` spends its entire budget, `small` will likely be left with some budget left
-/// to also do useful work. In particular, if the remaining budget was `N` at the start of `poll`,
-/// `small` will have at least a budget of `N - 64`. It may be more if `big` did not spend its
-/// entire budget.
-///
-/// Note that you cannot _increase_ your budget by calling `limit`. The budget provided to the code
-/// inside the buget is the _minimum_ of the _current_ budget and the bound.
-///
-#[allow(unreachable_pub, dead_code)]
-pub fn limit<R>(bound: usize, f: impl FnOnce() -> R) -> R {
```
Comment (jonhoo) [inline tokio/src/coop.rs:222]: Just for the record, I think it's going to be vital for us to expose something along these lines eventually.

Comment (carllerche) [reply tokio/src/coop.rs:222]: Fine w/ me. I just want to get real use cases documented & worked through before hand.

It should be easy enough to bring back once we need it.

**Pull request merged**